script/verification/HoloceneSystemConfigUpgrade.s.sol

```diff
@@ -0,0 +1,279 @@
+pragma solidity ^0.8.15;
+
+import {console2 as console} from "forge-std/console2.sol";
+import {LibString} from "solady/utils/LibString.sol";
+import {SuperchainRegistry} from "script/verification/Verification.s.sol";
+
+// HoloceneSystemConfigUpgrade is a contract that can be used to verify the Holocene upgrade of the SystemConfig contract.
+// The upgrade paths supported are:
+// 1.12.0 -> 2.3.0
+//  2.2.0 -> 2.3.0
+//  2.3.0 -> 2.3.0 (this case covers where the SystemConfig contract is already at the target version but was upgraded incorrectly)
+// The verification checks that the following storage values are properly modified if appropriate:
+// - scalar (may be migrated to a different encoding version)
+// - basefeeScalar (new storage variable, needs to be set in a way which is consistent with the previous scalar)
+// - blobbasefeeScalar (new storage variable, needs to be set in a way which is consistent with the previous scalar)
+// - gasPayingToken (new storage variable, needs to be set to a magic value representing ETH)
+// - disputeGameFactory (does not exist @ 1.12.0, needs to be set to zero address for changes on this upgrade path)
+// It also verified that remaining storage variables are unchanged.
+contract HoloceneSystemConfigUpgrade is SuperchainRegistry {
+    using LibString for string;
+
+    address public systemConfigAddress;
+    ISystemConfig sysCfg;
+
+    // Storage variables which exist in the existing and target
+    // SystemConfig contract versions irrespective of upgrade path,
+    // and which should be preserved:
+    struct BaseSysCfgVars {
+        address owner;
+        bytes32 batcherHash;
+        uint256 gasLimit;
+        address unsafeBlockSigner;
+        IResourceMetering.ResourceConfig resourceConfig;
+        address batchInbox;
+        address l1CrossDomainMessenger;
+        address l1StandardBridge;
+        address l1ERC721Bridge;
+        address optimismPortal;
+        address optimismMintableERC20Factory;
+    }
+
+    BaseSysCfgVars previous;
+
+    // Values which may change during the upgrade, depending on the chain::
+    uint256 previousScalar;
+
+    // The target value for the DisputeGameFactory address,
+    // this getter does not exist in the SystemConfig contract @ 1.12.0:
+    address targetDGF;
+
+    // Target values for gasPayingToken and its decimals. This should represent ETH
+    // on all upgrade paths:
+    address constant targetGasPayingToken = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+    uint8 constant targetDecimals = 18;
+
+    // The target version of the SystemConfig contract, read from the SCR @ specified release:
+    string targetVersion;
+
+    // The constructor caches some information from the Superchain Registry and from the existing SystemConfig contract
+    // before the upgrade is executed.
+    constructor(string memory _l1ChainName, string memory _l2ChainName, string memory _release)
+        SuperchainRegistry(_l1ChainName, _l2ChainName, _release)
```

Comment:
You can make the contract `abstract` and then don't need to pass down initialization of the `SuperchainRegistry` contract. This is more modular. See my recent fp upgrade tasks, where the actual check script instantiation directly calls the `SuperchainRegistry` constructor.

```suggestion
    constructor() # and make contract abstract
```

Reply:
Ahh I see that you need to make multiple instances and don't inherit this contract directly... ok then it makes sense.